File: tests/07_phases_hlt.dg/success_normalize_01.c
/*
<testinfo>
test_generator="config/mercurium-hlt run"
</testinfo>
*/

#include <stdio.h>
#include <stdlib.h>

char c[200];
char d[200];

void init()
{
    int i;
    for (i = 0; i < 200; i++)
    {
        c[i] = 0;
        d[i] = 0;
    }
}

void test1(void)
{
    init();

    int i;
    for (i = 4; i < 100; i+=7)
    {
        d[i]++;
    }

#pragma hlt normalize
    for (i = 4; i < 100; i+=7)
    {
        c[i]++;
    }

    for (i = 4; i < 100; i+=7)
    {
        if (c[i] != 1)
            abort();
    }

    for (i = 0; i < 200; i++)
    {
        if (d[i] != c[i])
            abort();
    }
}

void test2(void)
{

    init();

    int i;
    for (i = 100; i >= 3; i -= 11)
    {
        d[i]++;
    }

#pragma hlt normalize
    for (i = 100; i >= 3; i -= 11)
    {
        c[i]++;
    }

    for (i = 100; i >= 3; i -= 11)
    {
        if (c[i] != 1)
            abort();
    }

    for (i = 0; i < 200; i++)
    {
        if (d[i] != c[i])
            abort();
    }
}

int main(int argc, char *argv[])
{
    test1();
    test2();

    return 0;
}

File: tests/07_phases_hlt.dg/success_normalize_01.f90
! <testinfo>
! test_FFLAGS="--pp=on"
! test_generator="config/mercurium-hlt run"
! </testinfo>

SUBROUTINE INIT(N, V1, V2)
    IMPLICIT NONE
    INTEGER:: N
    INTEGER :: V1(N)
    INTEGER :: V2(N)

    V1 = 0
    V2 = 0
END SUBROUTINE INIT

PROGRAM P
    IMPLICIT NONE
    INTEGER, PARAMETER :: N = 100
    INTEGER :: V1(N)
    INTEGER :: V2(N)
    INTEGER :: I

    !-------------------------
    CALL INIT(N, V1, V2)
    DO I=1, N, 7
        V1(I) = V1(I) + 1
    ENDDO
    !$HLT NORMALIZE
    DO I=1, N, 7
        V2(I) = V2(I) + 1
    ENDDO
    !$HLT END NORMALIZE
    IF (ANY(V1 /= V2)) STOP 1
    !-------------------------


    !-------------------------
    CALL INIT(N, V1, V2)
    DO I=-100, -1, 4
        V1(I+101) = V1(I+101) + 1
    ENDDO
    !$HLT NORMALIZE
    DO I=-100, -1, 4
        V2(I+101) = V2(I+101) + 1
    ENDDO
    !$HLT END NORMALIZE
    IF (ANY(V1 /= V2)) STOP 1
    !-------------------------

END PROGRAM P

